### PR TITLE
Add GOVUK radios [part 8]

### PR DESCRIPTION
**Pull request by tombye**
Part 8 of migrating our radio buttons to use the [GOV.UK Frontend radios component](https://design-system.service.gov.uk/components/radios/).

https://www.pivotaltracker.com/story/show/170030262

Changes the radios on the /templates page used for adding and moving templates and folders.

Includes some slightly big changes to `NestedFieldMixin`, `GovukRadiosField` and `GovukCheckboxesField` to allow customisation of radios and checkboxes that render in a tree structure.

## Nested radios and checkboxes

It's useful to have some context on how nested fields work before reviewing this pull request.

The nested versions of radios and checkboxes work by:
1. converting `option`s (the WTForms class for a single radio/checkbox) to `item`s ([GOVUK Frontend data for a single radio/checkbox](https://design-system.service.gov.uk/components/radios/#options-inline-radios-example--items))
2. checking each `option` for any child `option`s and doing the same for each of them until no children are found (leaf node)
3. getting the HTML for this level of `option`s by prerendering them as a group of radios
4. doing this for each parent until you get to the root
5. render the root radios/checkboxes

For example, the end `option`s in these radios are 'Folder 13, Folder 16 and Folder 2' and their parent `option` is 'Folder 1':

tree_rendering_1

...so the `item`s for 'Folder 13, Folder 16 and Folder 2' will be prerendered and the result set as HTML below the 'Folder 1' `option`:

tree_rendering_2

This is then done for each grouping, until we hit the 'Templates' root:

tree_rendering_3

### CODE REVIEW

**Comment by benthorner** [comment]
This PR has been around for a while, do you still want to get this reviewed?

**Comment by tombye** [comment]
> This PR has been around for a while, do you still want to get this reviewed?

Yes please. I've rebased it against the latest master so should be ready to go again.

**Review by idavidmcdonald — COMMENTED** [review]
Here are a few initial comments but would definitely appreciate a walk through with you to go over some of the harder bits I haven't been able to get my head around please...

File: app/main/forms.py [hunk 1]
```diff
@@ -934,6 +934,10 @@ class GovukRadiosFieldWithNoneOption(FieldWithNoneOption, GovukRadiosField):
     pass
 
 
+class GovukRadiosFieldWithRequiredMessage(RadioFieldWithRequiredMessage, GovukRadiosField):
```
[the remark this answers is not in the document]
Comment (tombye) [reply app/main/forms.py:937]: Radios as plural (GovukRadiosField) comes from [the GOVUK Frontend radios component](https://design-system.service.gov.uk/components/radios) whereas the singular (RadioField) comes from [WTForms RadioField](https://wtforms.readthedocs.io/en/2.3.x/fields/#wtforms.fields.RadioField).
We discussed this outside of this PR and decided it should stay as is.

File: app/templates/forms/fields/radios/template.njk [hunk 2]
```diff
@@ -0,0 +1,128 @@
+{% from "components/error-message/macro.njk" import govukErrorMessage -%}
+{% from "components/fieldset/macro.njk" import govukFieldset %}
+{% from "components/hint/macro.njk" import govukHint %}
+{% from "components/label/macro.njk" import govukLabel %}
+
+
+{#- Copied from https://github.com/alphagov/govuk-frontend/blob/v2.13.0/src/components/radios/template.njk
```
[the remark this answers is not in the document]
Comment (tombye) [reply app/templates/forms/fields/radios/template.njk:7]: I agree, and we discussed this outside of this PR so am going to mark this as resolved.

File: app/main/forms.py [hunk 3]
```diff
@@ -880,6 +880,15 @@ def widget(self, field, param_extensions=None, **kwargs):
         return govuk_radios_field_widget(self, field, param_extensions=param_extensions, **kwargs)
 
 
+class GovukRadiosFieldWithNoneOption(FieldWithNoneOption, GovukRadiosField):
+    pass
+
+
+# NestedFieldMixin puts the items into a tree hierarchy, pre-rendering the sub-trees of the top-level items
+class GovukNestedRadiosField(NestedFieldMixin, GovukRadiosFieldWithNoneOption):
+    render_as_list = True
```
[the remark this answers is not in the document]
Comment (tombye) [reply app/main/forms.py:947]: I added some comments explaining this which, I think, are needed as the way nested fields render makes the use of `render_as_list` unclear: https://github.com/alphagov/notifications-admin/pull/3792/commits/be0291f8b3d4d7ed7a1ebd511b915d26b94fb03b

File: app/templates/forms/fields/radios/template.njk [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    Changes:
+    - `asList` option added the root `params` object to allow setting of the `.govuk-radios` and `.govuk-radios__item` element types
+    - Nunjucks-to-Jinja language problems fixed (see https://github.com/alphagov/govuk-frontend-jinja/blob/master/govuk_frontend_jinja/templates.py) -#}
```
[the remark this answers is not in the document]
Comment (tombye) [reply app/templates/forms/fields/radios/template.njk]: We discussed this outside of this PR and I rewrote these comments so the Nunjucks-to-Jinja language fixes are clearer: https://github.com/alphagov/notifications-admin/pull/3792/commits/46c0f7bdea2db0b455b5474d5ec7fb9620f99533

File: app/main/forms.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (idavidmcdonald) [reply app/main/forms.py:947]: So I think the comments you added are useful but I still don't think it's being passed through to the template.

The `govuk_checkboxes_field_widget` uses `render_as_list` as a variable in its code but
The `govuk_radios_field_widget` does not use `render_as_list` as a variable in its code. I think it is missing from line 817ish

**Review by idavidmcdonald — APPROVED** [review]
So I've spun this up locally and played with the forms to
- move templates between folders
- check which folders a team member should be given permission to see

They both appear OK

I've left two final comments that I think don't break the two forms above but are things that potentially do need tightening up for consistency/correctness

Thanks for improved comments in the last two commits, useful (as was chatting to you in person).

**Comment by tombye** [comment]
While this has been approved, I am closing it because:
- in discussions about the code changes, I believe the changes are quite hard to follow
- the difficulties people have had understanding the changes suggest there may be problems with the extensions I made to the WTForms APIs

I think the later reason needs some research so I'm going to organise that in order to get some assurances that the code is understandable before suggesting these changes again.
